ReactSkia/components/RSkComponentScrollView.cpp

```diff
@@ -0,0 +1,276 @@
+         case RNS_KEY_Up:
+             handled = setScrollOffset(SCROLL_PREV,containerHeight_,frame.size.height,scrollOffsetY_);
+         default: break;
+      }
+    }
+
+    if(surface_ != nullptr && handled) {
+       surface_->compositor()->begin();
+       if(layer() && layer()->type() == RnsShell::LAYER_TYPE_PICTURE)
+          static_cast<RnsShell::PictureLayer*>(layer().get())->setPicture(getPicture());
+       surface_->compositor()->commit();
+    } else {
+
+      /* TODO Element cannot handle event,so propogate the event to parent */
+
+    }
+
+    return handled;
+}
+
+inline bool RSkComponentScrollView::isInScrollArea(Rect childFrame,Rect parentFrame) {
+
+  if(!isHorizontalScroll() && (childFrame.origin.y >= scrollOffsetY_) && (childFrame.origin.y < (scrollOffsetY_ + parentFrame.size.height)))
+      return true;
+
+  if(isHorizontalScroll() && (childFrame.origin.x >= scrollOffsetX_) && (childFrame.origin.x < (scrollOffsetX_ + parentFrame.size.width)))
+      return true;
+
+  return false;
+}
+
+inline void RSkComponentScrollView::paintChild(Rect frame) {
+
+    SkCanvas scrollCanvas(scrollBitmap_);
+    std::vector<SkIRect> bitmapSurfaceDamage;
+    std::vector<std::shared_ptr<RSkComponent>> visibleChildList;
+
+    SkRect clipBound = SkRect::MakeEmpty();
+    PaintContext paintContext = {
+            &scrollCanvas,  // canvas
+            &bitmapSurfaceDamage, // damage rects
+            clipBound, // combined clip bounds from surfaceDamage
+            nullptr, // GrDirectContext
+    };
+    /* Prepaint */
+    for (auto &child : childList_) {
+      if(isInScrollArea(child->getComponentData().layoutMetrics.frame,frame)) {         static_cast<RnsShell::Layer*>(child->layer().get())->prePaint(paintContext);
+         visibleChildList.push_back(child);
+      }
+    }
+
+    clipBound = surface_->compositor()->beginClip(paintContext);
+
+    /* Paint */
+    for (auto &child : visibleChildList) {
+       static_cast<RnsShell::Layer*>(child->layer().get())->paint(paintContext);
+    }
+
+    bitmapSurfaceDamage.clear();
+}
+
+void RSkComponentScrollView::OnPaint(SkCanvas *canvas) {
+  auto component = getComponentData();
+  auto const &scrollViewProps = *std::static_pointer_cast<ScrollViewProps const>(component.props);
+
+  /*Retrieve Shadow Props*/
+  ShadowMetrics shadowMetrics{};
+  shadowMetrics.shadowColor=scrollViewProps.shadowColor;
+  shadowMetrics.shadowOffset=scrollViewProps.shadowOffset;
+  shadowMetrics.shadowOpacity=scrollViewProps.shadowOpacity;
+  shadowMetrics.shadowRadius=scrollViewProps.shadowRadius;
+
+  /* apply view style props */
+  auto borderMetrics= scrollViewProps.resolveBorderMetrics(component.layoutMetrics);
+
+  if(forceBitmapReset_) {
+     scrollBitmap_.reset();
+     forceBitmapReset_ = false;
+  }
+
+  Rect frame = component.layoutMetrics.frame;
+  if(childList_.size()) {
+     if(scrollBitmap_.drawsNothing()) {
+        scrollBitmap_.allocN32Pixels(containerWidth_,containerHeight_);
```

Comment:
When will this memory be freed

Reply:
The bitmap memory would be available until the scroll view component is deleted by Delete instruction from ShadowView Mutation.
SkBitmap has the ownership of the pointer which points to this pixel memory. This pointer is a shared pointer.When the RSKcomponentScrollView object is destroyed, SKBitmap object is destroyed as well which reduces ref count of the smart pointer  and memory associated with the same is released.